In [3]:
import numpy as np
import scipy.stats as stats
import pandas as pd
import csv 
from matplotlib import pyplot

In [4]:
DF = pd.read_csv("IPL_Twitter_MissingData.csv")#,header=None)

In [5]:
# print(DF.head())
# print(type(DF))

In [6]:
DF.isna().sum(axis = 1)
row_wise_nanvalues = pd.DataFrame(DF.isna().sum(axis = 1))
# print(row_wise_nanvalues[0] > 0)
row_missing_value = pd.DataFrame(row_wise_nanvalues[0] > 0) #like boolean in np boolean masking)
# print(row_missing_value)
DF1 =DF.copy()
DF2 = DF.copy()
# print(type(row_missing_value))
# DF1['row has nan?'] = row_missing_value[0]
# print(DF1.head())
# print(DF.head())

In [7]:
# series = pd.Series([True,False,True,False,False,False,True])
# series
# print(series.values)
# type(series.values)

In [8]:
nan_samples = np.sum(row_missing_value[0])
nan_samples
correct_samples = (DF2.shape)[0]-nan_samples
# print(type(nan_samples))
print('Number of samples with missing values are ',nan_samples, '.', sep ='' )
print('Number of samples with out missing values are ',correct_samples,'.', sep ='' )
print('As total number of samples are ',(DF2.shape)[0],'.', sep='')

Number of samples with missing values are 238.
Number of samples with out missing values are 762.
As total number of samples are 1000.


In [9]:
column_wise_nanvalues = pd.DataFrame(DF.isnull().sum(axis = 0))
type(column_wise_nanvalues) #data frame
column_wise_nanvalues.columns = ['number of values missing from each column']
column_wise_nanvalues

number of values missing from each column
Q1                                         44
Q2                                         45
X1                                         36
X2                                         55
X3                                         67
X4                                         41

## 1) How many data samples have missing values? 

if the column vector (Q1, Q2, X1, X2, X3, X4) represents a sample, then having at least one missing value in the sample includes it in missing values sample.

so, it is 238 samples with missing values and 1000-238 = 762 samples have complete data

In [10]:
# Df_without_nan = DF.where(DF['row has nan?'] == False)
# print(Df_without_nan.head())
# DF.shape

In [11]:
DF = DF.dropna()

In [12]:
print(DF.shape)
# DF.dropna().shape

(762, 6)


## 4. Impute the missing values using your own ideas and also briefly explain those ideas in 
the report (scores will be proportional to the number of meaningful ways the data is 
filled). One has to use one unique idea to fill the complete data so submit multiple sets 
of filled data, if you have multiple ideas for filling the dat## 4.

### filling it with a constant for Q1 and Q2 as they have to be binary

In [13]:
hhuh = pd.DataFrame([0,0,0,1,1,1,0,1,1,0,0,1,1,np.nan,np.nan,0,1,0,np.nan,1,1,0,1,0,1,np.nan,1,1])
# hhuh.interpolate(method ='linear', limit_direction ='backward', limit = 1) 
# df.interpolate?

In [14]:
# huh = pd.DataFrame([16,np.nan,36,0,1,4,9,49,64,np.nan,np.nan,121,144])
# huh.interpolate(method ='quadratic', limit_direction ='backward', limit = 1)  this won't work because the function is approximated as quadratic


In [15]:
#fill Q1 and Q2 by 1 if the number of 1 are more than 0's that is if (sum of entries in Q1/total entries after deleting nan entries in the column)
temp_col = DF2['Q1'].dropna()
checkQ1 = sum(temp_col)/temp_col.size
checkQ2 = sum(DF2['Q2'].dropna())/(DF2['Q1'].dropna().size)
# print(checkQ1,checkQ2)
if checkQ1>=0.5:
    DF2['Q1'].fillna(1)
else:
    DF2['Q1'] = DF2['Q1'].fillna(0)
#     pass
if checkQ2>=0.5:
    DF2['Q2'].fillna(1)
else:
    DF2['Q2'] = DF2['Q2'].fillna(0)
#     pass
# DF2

In [16]:
DF3 = pd.read_csv("IPL_Twitter_MissingData.csv")
DF4 = DF3.copy()
# DF3

### ffill and bfill

In [17]:
DF4[['Q1','Q2']] = DF4[['Q1','Q2']].bfill()
# DF4 #Q1 column 21 is filled with 0 (20 is 0, 21 is 1) and Q2 column 6 is filled with 1 (5,7 both are 1s)

In [18]:
DF3[['Q1','Q2']] = DF3[['Q1','Q2']].ffill() # here Q1 21 is filled by 1 as this is backward filling
DF7 = DF3.copy()

In [19]:
# DF4

### Filling other columns (integers)

fill by constant. By median or mean

In [20]:
# print(DF4['X1'].mean())
DF5 = DF2.copy()
DF6 = DF2.copy()
DF6 = DF6.interpolate(method='linear')
DF2['X1'] = DF2['X1'].fillna(int(DF2['X1'].mean()))
DF2['X2'] = DF2['X2'].fillna(int(DF2['X2'].mean()))
DF2['X3'] = DF2['X3'].fillna(int(DF2['X3'].mean()))
DF2['X4'] = DF2['X4'].fillna(int(DF2['X4'].mean()))
DF5['X1'] = DF5['X1'].fillna(int(DF5['X1'].median()))
DF5['X2'] = DF5['X2'].fillna(int(DF5['X2'].median()))
DF5['X3'] = DF5['X3'].fillna(int(DF5['X3'].median()))
DF5['X4'] = DF5['X4'].fillna(int(DF5['X4'].median()))
DF7['X1'] = DF7['X1'].fillna(int(DF7['X1'].mean()))
DF7['X2'] = DF7['X2'].fillna(int(DF7['X2'].mean()))
DF7['X3'] = DF7['X3'].fillna(int(DF7['X3'].mean()))
DF7['X4'] = DF7['X4'].fillna(int(DF7['X4'].mean()))

In [21]:
# DF5

In [22]:
# tempdf = pd.DataFrame([np.random.randint(1,100)+np.random.random(1) for i in range(10)])
# tempdf.round(1)

In [23]:
DF3[['X1','X2','X3','X4']] = DF3[['X1','X2','X3','X4']].ffill()
DF4[['X1','X2','X3','X4']] = DF4[['X1','X2','X3','X4']].bfill()

In [24]:
DF7 = DF7

In [25]:
# DF6

In [26]:
# DF3 == DF4
# DF2 == DF5

DF : dropped all rows with Nan values
Df2: threshold check, mean
Df3: bfill
Df4: ffill
Df5: threshold check, median
Df6: threshold check, linear interpolation
Df7 : ffill, mean

# 3. linear relationship between all things

In [27]:
# https://machinelearningmastery.com/how-to-use-correlation-to-understand-the-relationship-between-variables/
# np.cov?
# cov(X, Y) = E[(X - E[X] . (Y - E[Y])]
# cov(X, Y) = sum (x - E[X]) * (y - E[Y]) * 1/n
# Pearson's correlation coefficient = covariance(X, Y) / (stdv(X) * stdv(Y))
# stats.pearsonr?
# stats.spearmanr?
# Spearman's correlation coefficient = covariance(rank(X), rank(Y)) / (stdv(rank(X)) * stdv(rank(Y)))

after omitting rows with Nan values ,dataframe is DF.

filling nan values with ffill in binary case and using mean in the case of discrete variables is DF7

# a. For whole data

#### i . AFTER OMITTING NAN VALUES

In [40]:
# data1 = 20 * np.random.randn(1000) + 100
# np.shape(data1)
np.shape(Q1)

(762,)

In [63]:
print(len(DF.columns))
Q1 = DF['Q1'].values
Q2 = DF['Q2'].values
X1 = DF['X1'].values
X2 = DF['X2'].values
X3 = DF['X3'].values
X4 = DF['X4'].values
temp_cov = np.zeros([2*(len(DF.columns)),2*(len(DF.columns))])
temp_pearsonr= np.zeros([(len(DF.columns)),(len(DF.columns))])
temp_spearmanr= np.zeros([(len(DF.columns)),(len(DF.columns))])
# temp_cov[0:2,0:2]
DF.columns[0]

6


'Q1'

In [ ]:
# temp_cov = [[cov(Q1,Q1),cov(Q1,Q2),cove(Q1,X1).........],
#             [cov(Q2,Q1),cov(Q2,Q2),cov(Q2,Q3),.........],
#             .
#             .
#             .
#             .
#             [cov(X4,Q1),cov(X4,Q2),.....................]]
# # temp_pearsonr is similar

In [67]:
stats.pearsonr?

In [68]:

for i in range(len(DF.columns)):
    for j in range(len(DF.columns)):
        temp_cov[2*i:2*(i+1),2*(j):2*(1+j)] = np.cov(DF[DF.columns[i]],DF[DF.columns[j]])
        temp_pearsonr[i,j] = stats.pearsonr(DF[DF.columns[i]],DF[DF.columns[j]])[0] #first value of it is correlation coeff
        temp_spearmanr[i,j] = stats.spearmanr(DF[DF.columns[i]],DF[DF.columns[j]])[0]

In [69]:
np.shape(temp_cov)

(12, 12)

In [70]:
print(temp_cov)
print(temp_pearsonr)
print(temp_spearmanr)

[[ 2.50266434e-01  2.50266434e-01  2.50266434e-01  1.25456558e-02
   2.50266434e-01 -2.47572082e+02  2.50266434e-01  1.81174204e+03
   2.50266434e-01  6.60335096e+02  2.50266434e-01 -3.81405072e+02]
 [ 2.50266434e-01  2.50266434e-01  1.25456558e-02  2.50266434e-01
  -2.47572082e+02  3.54539231e+06  1.81174204e+03  1.74839477e+07
   6.60335096e+02  1.70395643e+07 -3.81405072e+02  9.05307647e+06]
 [ 2.50266434e-01  1.25456558e-02  2.50266434e-01  2.50266434e-01
   2.50266434e-01 -3.05351703e+02  2.50266434e-01  7.04288188e+02
   2.50266434e-01  1.77065702e+03  2.50266434e-01 -7.26774954e+02]
 [ 1.25456558e-02  2.50266434e-01  2.50266434e-01  2.50266434e-01
  -3.05351703e+02  3.54539231e+06  7.04288188e+02  1.74839477e+07
   1.77065702e+03  1.70395643e+07 -7.26774954e+02  9.05307647e+06]
 [ 3.54539231e+06 -2.47572082e+02  3.54539231e+06 -3.05351703e+02
   3.54539231e+06  3.54539231e+06  3.54539231e+06 -3.17773740e+06
   3.54539231e+06 -3.20202168e+06  3.54539231e+06  3.56752357e+06]
 [-2.

In [77]:
er = ((temp_pearsonr+temp_spearmanr)/2)>0.75 #strongly +vely linearly correlated ones wiht true

In [78]:
err = ((temp_pearsonr+temp_spearmanr)/2)<-0.75 #strongly negatively correlated

In [82]:
(((temp_pearsonr+temp_spearmanr)/2)>0.45) & (((temp_pearsonr+temp_spearmanr)/2)<0.75)  #fairly linearly correlated
# er & err

array([[False, False, False, False, False, False],
       [False, False, False, False, False, False],
       [False, False, False, False, False,  True],
       [False, False, False, False,  True, False],
       [False, False, False,  True, False, False],
       [False, False,  True, False, False, False]])

In [83]:
(((temp_pearsonr+temp_spearmanr)/2)<0.1) & (((temp_pearsonr+temp_spearmanr)/2)>-0.1) #barely(not at all) linearly correlated ones wiht true

array([[False,  True, False, False, False, False],
       [ True, False, False, False, False, False],
       [False, False, False, False, False, False],
       [False, False, False, False, False, False],
       [False, False, False, False, False, False],
       [False, False, False, False, False, False]])

In [84]:
(((temp_pearsonr+temp_spearmanr)/2)<-0.45) & (((temp_pearsonr+temp_spearmanr)/2)>-0.75)  ##fairly linearly negatively correlated

array([[False, False, False, False, False, False],
       [False, False, False, False, False,  True],
       [False, False, False, False, False, False],
       [False, False, False, False, False, False],
       [False, False, False, False, False,  True],
       [False,  True, False, False,  True, False]])

### ii. for filled Dataframe

In [85]:
Q1 = DF7['Q1'].values
Q2 = DF7['Q2'].values
X1 = DF7['X1'].values
X2 = DF7['X2'].values
X3 = DF7['X3'].values
X4 = DF7['X4'].values
temp_cov = np.zeros([2*(len(DF7.columns)),2*(len(DF7.columns))])
temp_pearsonr= np.zeros([(len(DF7.columns)),(len(DF7.columns))])
for i in range(len(DF7.columns)):
    for j in range(len(DF7.columns)):
        temp_cov[2*i:2*(i+1),2*(j):2*(1+j)] = np.cov(DF7[DF7.columns[i]],DF7[DF7.columns[j]])
        temp_pearsonr[i,j] = stats.pearsonr(DF7[DF7.columns[i]],DF7[DF7.columns[j]])[0] #first value of it is correlation coeff
        temp_spearmanr[i,j] = stats.spearmanr(DF7[DF7.columns[i]],DF7[DF7.columns[j]])[0]

# B. IF THERE IS NO MATCH FOR BOTH CSK AND MI

In [86]:
## for filled datframe

In [92]:
Df = DF.where((DF['Q1'] == 0) & (DF['Q2'] == 0)).dropna() #similar for other things

In [95]:
Df.shape #200 matches where none of both played
DF = Df.copy()

In [96]:
Q1 = DF['Q1'].values
Q2 = DF['Q2'].values
X1 = DF['X1'].values
X2 = DF['X2'].values
X3 = DF['X3'].values
X4 = DF['X4'].values
temp_cov = np.zeros([2*(len(DF.columns)),2*(len(DF.columns))])
temp_pearsonr= np.zeros([(len(DF.columns)),(len(DF.columns))])
temp_spearmanr= np.zeros([(len(DF.columns)),(len(DF.columns))])
for i in range(len(DF.columns)):
    for j in range(len(DF.columns)):
        temp_cov[2*i:2*(i+1),2*(j):2*(1+j)] = np.cov(DF[DF.columns[i]],DF[DF.columns[j]])
        temp_pearsonr[i,j] = stats.pearsonr(DF[DF.columns[i]],DF[DF.columns[j]])[0] #first value of it is correlation coeff
        temp_spearmanr[i,j] = stats.spearmanr(DF[DF.columns[i]],DF[DF.columns[j]])[0]

C:\Users\sudhe\Anaconda3\lib\site-packages\scipy\stats\stats.py:3010: RuntimeWarning: invalid value encountered in double_scalars
  r = r_num / r_den
C:\Users\sudhe\Anaconda3\lib\site-packages\scipy\stats\stats.py:5256: RuntimeWarning: invalid value encountered in less
  x = np.where(x < 1.0, x, 1.0)  # if x > 1 then return 1.0
C:\Users\sudhe\Anaconda3\lib\site-packages\numpy\lib\function_base.py:2530: RuntimeWarning: invalid value encountered in true_divide
  c /= stddev[:, None]
C:\Users\sudhe\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in greater
  return (self.a < x) & (x < self.b)
C:\Users\sudhe\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:879: RuntimeWarning: invalid value encountered in less
  return (self.a < x) & (x < self.b)
C:\Users\sudhe\Anaconda3\lib\site-packages\scipy\stats\_distn_infrastructure.py:1821: RuntimeWarning: invalid value encountered in less_equal
  cond2 = cond0 & (x <= s

In [97]:
print('\n strongly +vely linearly correlated ones wiht true \n')
print(((temp_pearsonr+temp_spearmanr)/2)>0.75) #strongly +vely linearly correlated ones wiht true
print('\n strongly negatively correlated \n')
print(((temp_pearsonr+temp_spearmanr)/2)<-0.75)
 #strongly negatively correlated
print('\n fairly linearly correlated \n')
print((((temp_pearsonr+temp_spearmanr)/2)>0.45) & (((temp_pearsonr+temp_spearmanr)/2)<0.75))
print('\n barely(not at all) linearly correlated ones wiht true \n')
print((((temp_pearsonr+temp_spearmanr)/2)<0.1) & (((temp_pearsonr+temp_spearmanr)/2)>-0.1))
print('\n fairly linearly negatively correlated \n')
print((((temp_pearsonr+temp_spearmanr)/2)<-0.45) & (((temp_pearsonr+temp_spearmanr)/2)>-0.75))
  ##




 strongly +vely linearly correlated ones wiht true 

[[False False False False False False]
 [False False False False False False]
 [False False  True False False  True]
 [False False False  True False False]
 [False False False False  True False]
 [False False  True False False  True]]

 strongly negatively correlated 

[[False False False False False False]
 [False False False False False False]
 [False False False False False False]
 [False False False False False False]
 [False False False False False False]
 [False False False False False False]]

 fairly linearly correlated 

[[False False False False False False]
 [False False False False False False]
 [False False False False False False]
 [False False False False False  True]
 [False False False False False False]
 [False False False  True False False]]

 barely(not at all) linearly correlated ones wiht true 

[[False False False False False False]
 [False False False False False False]
 [False False False  True False False]


C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:2: RuntimeWarning: invalid value encountered in greater
  
C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:4: RuntimeWarning: invalid value encountered in less
  after removing the cwd from sys.path.
C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: RuntimeWarning: invalid value encountered in greater
  import sys
C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:7: RuntimeWarning: invalid value encountered in less
  import sys
C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: RuntimeWarning: invalid value encountered in less
  if __name__ == '__main__':
C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:9: RuntimeWarning: invalid value encountered in greater
  if __name__ == '__main__':
C:\Users\sudhe\Anaconda3\lib\site-packages\ipykernel_launcher.py:11: RuntimeWarning: invalid value encountered in less
  # This is added back by Interactive